### ISSUE TITLE
Explore "BLoC"

### ISSUE THREAD

**Issue by tenhobi** [tenhobi/flashcards#12]
In [this video](https://www.youtube.com/watch?v=PLHln7wHgPE) from the DartConf, one of the spokesmans was saying about this "BLoC" – Bussiness Logic Component, which has really pedantic rules to follow, but provides us a nice way how to separate a logic of an app from views (angular, flutter).

---

BLoC designing guidelines:

1. Inputs and outputs are simple Streams/Sinks only
1. Dependencies must be injectable and platform agnostic
1. No platform branching allowed
1. Implementation can be whatever you want if you follow the previous rules

UI design guidelines:

1. Each "complex enough" component has a corresponding BLoC
1. Components should send inputs "as is"
1. Components should show outputs as close as possible to "as is"
1. All branching should be based on simple BLoC boolean outputs

TL;DR

1. Move bussiness logic to BLoCs
1. Keep UI components simple

---

The task is to explore the video, try to underestand the concept and make some example app with this.

**Comment by tenhobi** [comment]
Because our source is a video and we do not wanna go through it over and over again, I did some screenshots as shown below.

Basically, to get this to work, we "only" have to separate UIs and business logic. If some platform dependent API is needed, it should be passed as a parameter and the API should be derivation of API's interface.

On the UI side, all inputs and outputs are provided through the business logic. Because Flutter app needs to be reloaded on each change in order to redraw, it's a good idea to use `StreamBulder`, in Angular, there is `| async` pipe.

The business logic therefore uses inputs (`Sink`), works with (platform dependent) APIs and generates some outputs (`Stream`).

---

*Image 1*
![screenshot_1](https://user-images.githubusercontent.com/5287596/37556302-6738673c-29f4-11e8-9455-397cb8f66d4c.png)

*Image 2*
![screenshot_2](https://user-images.githubusercontent.com/5287596/37556304-67f7cbd6-29f4-11e8-898f-cca3a6417cd8.png)

*Image 3*
![screenshot_3](https://user-images.githubusercontent.com/5287596/37556307-69f877d2-29f4-11e8-9be2-4dc6168291d2.png)

*Image 4*
![screenshot_4](https://user-images.githubusercontent.com/5287596/37556308-6a30b76e-29f4-11e8-8c80-8eecdd78b6f6.png)

*Image 5*
![screenshot_5](https://user-images.githubusercontent.com/5287596/37556309-6a4f7df2-29f4-11e8-9416-eb8209f31e01.png)

*Image 6*
![screenshot_6](https://user-images.githubusercontent.com/5287596/37556306-69dd7cca-29f4-11e8-8640-445f441b6afd.png)

To ensure everything will work good, the developers must strictly follow **"aggressively pedantic rules"**:

1. Inputs and outputs are simple Streams/Sinks only
1. Dependencies must be injectable and platform agnostic
1. No platform branching allowed
1. Implementation can be whatever you want if you follow the previous rules

And of course, not everything has to have its own BLoC, only "complex enough" component should have one:

1. Each "complex enough" component has a corresponding BLoC
1. Components should send inputs "as is"
1. Components should show outputs as close as possible to "as is"
1. All branching should be based on simple BLoC boolean outputs

TL;DR:

1. Move bussiness logic to BLoCs
1. Keep UI components simple
1. **Design rules aren't negotiable**

---

I think this is a quite good way and it worth a try. As an output of this project, we may do some review of this BLoC.
